File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/catalog/sets/SetListFragment.java
package com.lucagiorgetti.surprix.ui.mainfragments.catalog.sets;

import android.app.AlertDialog;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.SearchView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.google.android.material.snackbar.Snackbar;
import com.lucagiorgetti.surprix.R;
import com.lucagiorgetti.surprix.SurprixApplication;
import com.lucagiorgetti.surprix.model.Set;
import com.lucagiorgetti.surprix.ui.mainfragments.catalog.CatalogNavigationMode;
import com.lucagiorgetti.surprix.ui.mainfragments.filter.ChipFilters;
import com.lucagiorgetti.surprix.ui.mainfragments.filter.CollectionSetFilterBSDFragment;
import com.lucagiorgetti.surprix.ui.mainfragments.filter.FilterBottomSheetListener;
import com.lucagiorgetti.surprix.utility.SystemUtils;
import com.lucagiorgetti.surprix.utility.dao.CollectionDao;
import com.lucagiorgetti.surprix.utility.dao.MissingListDao;

public class SetListFragment extends BaseSetListFragment {
    CatalogNavigationMode navigationMode;
    String yearId;
    String producerId;
    CollectionDao collectionDao;
    MissingListDao missingListDao;
    SetListViewModel setListViewModel;
    private ChipFilters chipFilters;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        String username = SurprixApplication.getInstance().getCurrentUser().getUsername();
        collectionDao = new CollectionDao(username);
        missingListDao = new MissingListDao(username);

        if (getArguments() != null) {
            yearId = SetListFragmentArgs.fromBundle(getArguments()).getYearId();
            producerId = SetListFragmentArgs.fromBundle(getArguments()).getProducerId();
            String yearName = SetListFragmentArgs.fromBundle(getArguments()).getYearName();
            navigationMode = SetListFragmentArgs.fromBundle(getArguments()).getNavigationMode();
            setTitle(yearName);

            if (navigationMode.equals(CatalogNavigationMode.CATALOG) && !SystemUtils.isSetHintDisplayed()) {
                showHintAlert();
            }
        }

        super.onCreate(savedInstanceState);
    }

    private void showHintAlert() {
        androidx.appcompat.app.AlertDialog dialog = new androidx.appcompat.app.AlertDialog.Builder(getContext())
                .setTitle("Hint")
                .setMessage("To easily create your missing list, add the set to your collection and keep the set pressed. You will add all of its surprises to your missing list!")
                .setPositiveButton("Ok. thanks", (dialog1, which) -> {
                    SystemUtils.setSetHintDisplayed(true);
                    dialog1.dismiss();
                })
                .create();

        dialog.setCanceledOnTouchOutside(false);
        dialog.show();
    }

    @Override
    public void setupView() {
        setListViewModel = new ViewModelProvider(this).get(SetListViewModel.class);
        setListViewModel.getSets(yearId, producerId, navigationMode).observe(getViewLifecycleOwner(), sets -> {
            mAdapter.submitList(sets);
            mAdapter.setFilterableList(sets);

            if (navigationMode == CatalogNavigationMode.COLLECTION) {
                chipFilters = new ChipFilters();
                chipFilters.initByCatalogSets(sets);
            }
        });

        setListViewModel.isLoading().observe(getViewLifecycleOwner(), isLoading -> progress.setVisibility(isLoading ? View.VISIBLE : View.GONE));
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        switch (navigationMode) {
            case CATALOG:
                inflater.inflate(R.menu.search_menu, menu);
                break;
            case COLLECTION:
                inflater.inflate(R.menu.filter_search_menu, menu);
                break;
        }

        MenuItem searchItem = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) searchItem.getActionView();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                mAdapter.getFilter().filter(s);
                return false;
            }
        });
        searchView.setQueryHint(getString(R.string.search));
        searchView.setOnCloseListener(() -> {
            if (mAdapter.getItemCount() > 0) {
                setTitle(getString(R.string.missings) + " (" + mAdapter.getItemCount() + ")");
            } else {
                setTitle(getString(R.string.missings));
            }
            return false;
        });

        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_filter) {
            openBottomSheet();
        }
        return super.onOptionsItemSelected(item);
    }

    private void openBottomSheet() {
        if (chipFilters != null) {
            CollectionSetFilterBSDFragment bottomSheetDialogFragment = new CollectionSetFilterBSDFragment(this.chipFilters);
            bottomSheetDialogFragment.setListener(new FilterBottomSheetListener() {
                @Override
                public void onFilterChanged(ChipFilters selection) {
                    mAdapter.setChipFilters(selection);
                }

                @Override
                public void onFilterCleared() {
                    chipFilters.clearSelection();
                    mAdapter.setChipFilters(chipFilters);
                    bottomSheetDialogFragment.dismissAllowingStateLoss();
                }
            });
            bottomSheetDialogFragment.show(getActivity().getSupportFragmentManager(), "double_bottom_sheet");
        } else {
            Snackbar.make(getView(), SurprixApplication.getInstance().getString(R.string.cannot_oper_filters), Snackbar.LENGTH_SHORT).show();
        }
    }

    @Override
    public SetRecyclerAdapter setAdapter() {
        return new SetRecyclerAdapter(navigationMode, new MyClickListener() {
            @Override
            public void onSetInCollectionChanged(Set set, boolean isChecked) {
                if (isChecked) {
                    collectionDao.addSetInCollection(set);
                } else {
                    alertRemoveCollection(set);
                }
            }

            @Override
            public void onSetClicked(Set set) {
                SetListFragmentDirections.SetSelectedAction action = SetListFragmentDirections.setSelectedAction(set.getId(), set.getName(), navigationMode);
                Navigation.findNavController(getView()).navigate(action);
                SystemUtils.closeKeyboard(getActivity());
            }

            @Override
            public boolean onSetLongClicked(Set set, boolean inCollection) {
                onLongSetClicked(set, inCollection);
                return true;
            }
        });
    }

    private void alertRemoveCollection(Set set) {
        final androidx.appcompat.app.AlertDialog alertDialog = new androidx.appcompat.app.AlertDialog.Builder(getActivity()).create();
        alertDialog.setTitle(getString(R.string.remove_from_collection_title));
        alertDialog.setMessage(getString(R.string.remove_from_collection_message));
        alertDialog.setButton(androidx.appcompat.app.AlertDialog.BUTTON_POSITIVE, getString(R.string.proceed_btn),
                (dialog, which) -> {
                    missingListDao.removeItemsFromSet(set);
                    collectionDao.removeSetInCollection(set);
                    alertDialog.dismiss();
                });
        alertDialog.setButton(androidx.appcompat.app.AlertDialog.BUTTON_NEGATIVE, getString(R.string.discard_btn),
                (dialog, which) -> {
                    alertDialog.dismiss();
                    setListViewModel.getSets(yearId, producerId, navigationMode);
                    mAdapter.notifyDataSetChanged();
                });
        alertDialog.show();
    }

    private void onLongSetClicked(Set set, boolean inCollection) {
        final AlertDialog alertDialog = new AlertDialog.Builder(getActivity()).create();
        alertDialog.setTitle(getString(R.string.dialog_add_set_title));
        alertDialog.setMessage(getString(R.string.dialog_add_set_text) + " " + set.getName() + "?");
        alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, getString(R.string.dialog_positive),
                (dialog, which) -> {
                    new MissingListDao(SurprixApplication.getInstance().getCurrentUser().getUsername()).addMissingsBySet(set.getId());

                    if (!inCollection) {
                        setListViewModel.loadSets(yearId, producerId, navigationMode);
                    }
                    alertDialog.dismiss();
                });
        alertDialog.setButton(AlertDialog.BUTTON_NEGATIVE, getString(R.string.discard_btn),
                (dialog, which) -> alertDialog.dismiss());
        alertDialog.show();
    }

    public interface MyClickListener {
        void onSetInCollectionChanged(Set set, boolean isChecked);

        void onSetClicked(Set set);

        boolean onSetLongClicked(Set set, boolean inCollection);
    }

}

File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/catalog/sets/SetListFragment.kt
package com.lucagiorgetti.surprix.ui.mainfragments.catalog.sets

import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.SearchView
import androidx.appcompat.app.AlertDialog
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.snackbar.Snackbar
import com.lucagiorgetti.surprix.R
import com.lucagiorgetti.surprix.SurprixApplication
import com.lucagiorgetti.surprix.model.Set
import com.lucagiorgetti.surprix.utility.BaseFragment
import com.lucagiorgetti.surprix.utility.SystemUtils
import com.lucagiorgetti.surprix.utility.dao.MissingListDao

class SetListFragment : BaseFragment() {
    private var yearId: String? = null
    private var yearName: String? = null
    private var producerId: String? = null
    private var missingListDao: MissingListDao? = null
    private var setListViewModel: SetListViewModel? = null
    private var mAdapter: SetRecyclerAdapter? = null
    private var recyclerView: RecyclerView? = null
    private var swipeRefreshLayout: SwipeRefreshLayout? = null

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_set_list, container, false)
        mAdapter = setAdapter()
        val progress = root.findViewById<ProgressBar>(R.id.set_loading)
        setProgressBar(progress)
        recyclerView = root.findViewById(R.id.set_recycler)
        recyclerView!!.setHasFixedSize(true)
        val layoutManager: RecyclerView.LayoutManager = GridLayoutManager(context, SystemUtils.getColumnsNumber(recyclerView))
        recyclerView!!.layoutManager = layoutManager
        recyclerView!!.adapter = mAdapter
        swipeRefreshLayout = root.findViewById(R.id.swipe_refresh)
        swipeRefreshLayout!!.setOnRefreshListener { reloadData() }

        setupView()
        return root
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        val username = SurprixApplication.instance.currentUser?.username

        missingListDao = MissingListDao(username)
        if (arguments != null) {
            yearId = SetListFragmentArgs.fromBundle(requireArguments()).yearId
            producerId = SetListFragmentArgs.fromBundle(requireArguments()).producerId
            yearName = SetListFragmentArgs.fromBundle(requireArguments()).yearName
        }

        if (!SystemUtils.getHintDisplayed()) {
            showHintAlert()
        }

        super.onCreate(savedInstanceState)
    }

    private fun showHintAlert() {
        val dialog = AlertDialog.Builder(requireContext())
                .setTitle(R.string.set_hint_dialog_title)
                .setMessage(R.string.set_hint_dialog_message)
                .setPositiveButton(R.string.btn_ok_thanks) { dialog1: DialogInterface, _: Int ->
                    SystemUtils.setHintDisplayed(true)
                    dialog1.dismiss()
                }
                .create()
        dialog.setCanceledOnTouchOutside(false)
        dialog.show()
    }

    private fun reloadData() {
        setListViewModel!!.loadSets(yearId)
    }

    private fun setupView() {

        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                // Add menu items here
                menuInflater.inflate(R.menu.search_menu, menu)

                val searchItem = menu.findItem(R.id.action_search)
                val searchView = searchItem.actionView as SearchView?
                searchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                    override fun onQueryTextSubmit(s: String): Boolean {
                        return false
                    }

                    override fun onQueryTextChange(s: String): Boolean {
                        mAdapter!!.filter.filter(s)
                        return false
                    }
                })
                searchView.queryHint = getString(R.string.search)
                searchView.setOnCloseListener {
                    if (mAdapter!!.itemCount > 0) {
                        setTitle(getString(R.string.missings) + " (" + mAdapter!!.itemCount + ")")
                    } else {
                        setTitle(getString(R.string.missings))
                    }
                    false
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return false
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)

        setTitle(yearName!!)

        setListViewModel = ViewModelProvider(this)[SetListViewModel::class.java]
        setListViewModel!!.getSets(yearId).observe(viewLifecycleOwner) { sets: List<Set> ->
            mAdapter!!.submitList(sets)
            mAdapter!!.setFilterableList(sets)
            mAdapter!!.notifyDataSetChanged()
        }

        setListViewModel!!.isLoading.observe(viewLifecycleOwner) { isLoading: Boolean ->
            if (isLoading) {
                showLoading()
            } else {
                hideLoading()
                swipeRefreshLayout!!.isRefreshing = false
            }
        }
    }

    private fun setAdapter(): SetRecyclerAdapter {
        return SetRecyclerAdapter(object : MyClickListener {

            override fun onSetClicked(set: Set) {
                val thanks = set.thanks_to?.split(",")?.toTypedArray()
                val action = SetListFragmentDirections.setSelectedAction(set.id!!, set.name!!, thanks)
                findNavController(view!!).navigate(action)
                SystemUtils.closeKeyboard(activity)
            }

            override fun onSetLongClicked(set: Set): Boolean {
                onLongSetClicked(set)
                return true
            }
        })
    }

    private fun onLongSetClicked(set: Set) {
        val alertDialog = android.app.AlertDialog.Builder(activity).create()
        alertDialog.setTitle(getString(R.string.dialog_add_set_title))
        alertDialog.setMessage(getString(R.string.dialog_add_set_text) + " " + set.name + "?")
        alertDialog.setButton(android.app.AlertDialog.BUTTON_POSITIVE, getString(R.string.dialog_positive)
        ) { _: DialogInterface?, _: Int ->
            MissingListDao(SurprixApplication.instance.currentUser?.username).addMissingsBySet(set.id)
            alertDialog.dismiss()
            Snackbar.make(requireView(), SurprixApplication.instance.getString(R.string.added_to_missings), Snackbar.LENGTH_SHORT).show()
        }
        alertDialog.setButton(android.app.AlertDialog.BUTTON_NEGATIVE, getString(R.string.discard_btn)
        ) { _: DialogInterface?, _: Int -> alertDialog.dismiss() }
        alertDialog.show()
    }

    interface MyClickListener {
        fun onSetClicked(set: Set)

        fun onSetLongClicked(set: Set): Boolean
    }
}